File: sdks/node/src/webhooks.ts
import * as crypto from 'crypto';
import { WebhookEvent } from './types';
import { HookflowError } from './errors';

const SIGNATURE_HEADER = 'x-signature';
const TIMESTAMP_HEADER = 'x-timestamp';
const EVENT_ID_HEADER = 'x-event-id';
const DELIVERY_ID_HEADER = 'x-delivery-id';

const DEFAULT_TOLERANCE = 300000; // 5 minutes in milliseconds

export interface WebhookHeaders {
  'x-signature'?: string;
  'x-timestamp'?: string;
  'x-event-id'?: string;
  'x-delivery-id'?: string;
  [key: string]: string | undefined;
}

export interface VerifyOptions {
  tolerance?: number;
}

/**
 * Verifies the webhook signature using HMAC-SHA256.
 *
 * The header is `t=<unix-ms>,v1=<hex>`, and it may carry **more than one** `v1`.
 * After you rotate an endpoint's secret, Hookflow signs each delivery with both
 * the new secret and the retired one for the endpoint's grace window (24 hours
 * by default), so you can deploy the new secret whenever you like instead of at
 * the instant you press rotate. The delivery is authentic if *any* `v1` matches,
 * which is what this checks.
 *
 * @param payload - Raw request body as string
 * @param signature - X-Signature header value (format: t=timestamp,v1=signature[,v1=...])
 * @param secret - Endpoint webhook secret
 * @param options - Verification options
 * @returns true if signature is valid
 * @throws HookflowError if signature is invalid
 */
export function verifySignature(
  payload: string,
  signature: string,
  secret: string,
  options: VerifyOptions = {}
): boolean {
  const tolerance = options.tolerance ?? DEFAULT_TOLERANCE;

  if (!signature) {
    throw new HookflowError('Missing signature header', 400, 'invalid_signature');
  }

  const parts = signature.split(',');
  let timestamp: string | undefined;
  const signatures: string[] = [];

  for (const part of parts) {
    const [key, value] = part.split('=');
    if (key === 't') timestamp = value?.trim();
    // Collected, not overwritten: a header sent during a secret rotation carries
    // one v1 per valid secret, and keeping only the last would reject whichever
    // of the pair you are currently holding.
    if (key === 'v1' && value) signatures.push(value.trim());
  }

  if (!timestamp || signatures.length === 0) {
    throw new HookflowError(
      'Invalid signature format. Expected: t=timestamp,v1=signature',
      400,
      'invalid_signature'
    );
  }

  const timestampMs = parseInt(timestamp, 10);
  const now = Date.now();

  if (Math.abs(now - timestampMs) > tolerance) {
    throw new HookflowError(
      'Webhook timestamp is outside tolerance window',
      400,
      'timestamp_expired'
    );
  }

  const signedPayload = `${timestamp}.${payload}`;
  const expectedSignature = crypto
    .createHmac('sha256', secret)
    .update(signedPayload)
    .digest('hex');

  const expectedBuffer = Buffer.from(expectedSignature);
  // Every candidate is compared — no early exit — so the time taken does not
  // depend on which position matched.
  let matched = false;
  for (const candidate of signatures) {
    const sigBuffer = Buffer.from(candidate);
    if (sigBuffer.length === expectedBuffer.length && crypto.timingSafeEqual(sigBuffer, expectedBuffer)) {
      matched = true;
    }
  }

  if (!matched) {
    throw new HookflowError('Invalid signature', 400, 'invalid_signature');
  }

  return true;
}

const STANDARD_ID_HEADER = 'webhook-id';
const STANDARD_TIMESTAMP_HEADER = 'webhook-timestamp';
const STANDARD_SIGNATURE_HEADER = 'webhook-signature';

/** Tolerance for the Standard Webhooks headers, whose timestamp is in **seconds**. */
const DEFAULT_STANDARD_TOLERANCE_SECONDS = 300;

/**
 * Verifies the [Standard Webhooks](https://www.standardwebhooks.com) headers.
 *
 * Endpoints receive both header sets by default (`signatureScheme: 'BOTH'`), so use
 * whichever you prefer — this one if you would rather your verification match what other
 * providers send, `verifySignature` if you are already verifying `X-Signature`.
 *
 * Two things differ from Hookflow's own scheme beyond the header names: the message id is
 * part of what is signed, and the digest is base64 rather than hex. Rotation works the same
 * way — during the grace window the header carries a space-separated signature per valid
 * secret, and any one matching is enough.
 *
 * @param payload - Raw request body as string
 * @param headers - Request headers, including the three `webhook-*` ones
 * @param secret - The endpoint's `standardWebhooksSecret` (`whsec_…`), not the raw secret.
 *                 A plain secret is accepted too and used as-is.
 * @returns true if the signature is valid
 * @throws HookflowError if it is not
 */
export function verifyStandardWebhook(
  payload: string,
  headers: WebhookHeaders,
  secret: string,
  options: { toleranceSeconds?: number } = {}
): boolean {
  const tolerance = options.toleranceSeconds ?? DEFAULT_STANDARD_TOLERANCE_SECONDS;

  const id = headers[STANDARD_ID_HEADER] || headers['Webhook-Id'];
  const timestamp = headers[STANDARD_TIMESTAMP_HEADER] || headers['Webhook-Timestamp'];
  const signature = headers[STANDARD_SIGNATURE_HEADER] || headers['Webhook-Signature'];

  if (!id || !timestamp || !signature) {
    throw new HookflowError(
      'Missing webhook-id, webhook-timestamp or webhook-signature header',
      400,
      'invalid_signature'
    );
  }

  const timestampSeconds = parseInt(timestamp, 10);
  if (Number.isNaN(timestampSeconds)) {
    throw new HookflowError('Invalid webhook-timestamp header', 400, 'invalid_signature');
  }

  const nowSeconds = Math.floor(Date.now() / 1000);
  if (Math.abs(nowSeconds - timestampSeconds) > tolerance) {
    throw new HookflowError(
      'Webhook timestamp is outside tolerance window',
      400,
      'timestamp_expired'
    );
  }

  // `whsec_<base64>` is the conventional form and is what the endpoint's
  // standardWebhooksSecret gives you; the base64 body decodes to the key bytes. Anything
  // else is taken literally, so a raw secret still works.
  const key = secret.startsWith('whsec_')
    ? Buffer.from(secret.slice('whsec_'.length), 'base64')
    : Buffer.from(secret, 'utf8');

  const expected = crypto
    .createHmac('sha256', key)
    .update(`${id}.${timestampSeconds}.${payload}`)
    .digest('base64');
  const expectedBuffer = Buffer.from(expected);

  // Space-separated, one per valid secret during a rotation window. Every candidate is
  // compared with no early exit, so the time taken does not reveal which one matched.
  let matched = false;
  for (const part of signature.trim().split(/\s+/)) {
    const comma = part.indexOf(',');
    if (comma < 1 || part.slice(0, comma) !== 'v1') continue;
    const candidate = Buffer.from(part.slice(comma + 1));
    if (candidate.length === expectedBuffer.length && crypto.timingSafeEqual(candidate, expectedBuffer)) {
      matched = true;
    }
  }

  if (!matched) {
    throw new HookflowError('Invalid signature', 400, 'invalid_signature');
  }

  return true;
}

/**
 * Constructs a webhook event from the request.
 *
 * What Hookflow actually PUTs on the wire is the event's **payload**, not an
 * envelope: a `client.events.send({ type: 'order.completed', data: {...} })`
 * arrives at your endpoint as the `data` object alone, with the identifiers
 * carried in headers (`X-Event-Id`, `X-Delivery-Id`, `X-Timestamp`,
 * `X-Sequence-Number`). So:
 *
 * - `eventId` / `deliveryId` / `timestamp` come from the headers and are
 *   always populated for a real delivery.
 * - `data` is the parsed body.
 * - `type` is only populated when the body happens to carry a `type` key —
 *   which for a default subscription it does not. Route on the payload, or
 *   configure the subscription's `payloadTemplate` to wrap the event so that
 *   `type` is part of the body.
 *
 * @param payload - Raw request body as string
 * @param headers - Request headers
 * @param secret - Endpoint webhook secret
 * @param options - Verification options
 * @returns Parsed and verified webhook event
 */
export function constructEvent(
  payload: string,
  headers: WebhookHeaders,
  secret: string,
  options: VerifyOptions = {}
): WebhookEvent {
  const signature = headers[SIGNATURE_HEADER] || headers['X-Signature'];
  const timestamp = headers[TIMESTAMP_HEADER] || headers['X-Timestamp'];
  const eventId = headers[EVENT_ID_HEADER] || headers['X-Event-Id'];
  const deliveryId = headers[DELIVERY_ID_HEADER] || headers['X-Delivery-Id'];

  if (!signature) {
    throw new HookflowError('Missing X-Signature header', 400, 'missing_header');
  }

  verifySignature(payload, signature, secret, options);

  let data: Record<string, unknown>;
  try {
    data = JSON.parse(payload);
  } catch {
    throw new HookflowError('Invalid JSON payload', 400, 'invalid_payload');
  }

  return {
    eventId: eventId || '',
    deliveryId: deliveryId || '',
    timestamp: timestamp ? parseInt(timestamp, 10) : Date.now(),
    type: (data.type as string) || '',
    data: (data.data as Record<string, unknown>) || data,
  };
}

/**
 * Generates a signature for testing purposes
 * @param payload - Request body as string
 * @param secret - Webhook secret
 * @param timestamp - Optional timestamp (defaults to now)
 * @returns Signature string in format t=timestamp,v1=signature
 */
export function generateSignature(
  payload: string,
  secret: string,
  timestamp?: number
): string {
  const ts = timestamp ?? Date.now();
  const signedPayload = `${ts}.${payload}`;
  const signature = crypto
    .createHmac('sha256', secret)
    .update(signedPayload)
    .digest('hex');

  return `t=${ts},v1=${signature}`;
}

File: sdks/python/hookflow/webhooks.py
"""Webhook signature verification utilities."""

import base64
import hashlib
import hmac
import time
from typing import Any, Dict, Optional

from .types import WebhookEvent
from .errors import HookflowError

DEFAULT_TOLERANCE_MS = 300000  # 5 minutes

# The Standard Webhooks headers carry seconds, not milliseconds.
DEFAULT_STANDARD_TOLERANCE_SECONDS = 300


def verify_signature(
    payload: str,
    signature: str,
    secret: str,
    tolerance_ms: int = DEFAULT_TOLERANCE_MS,
) -> bool:
    """
    Verify webhook signature using HMAC-SHA256.

    The header is ``t=<unix-ms>,v1=<hex>`` and may carry more than one ``v1``.
    After you rotate an endpoint's secret, Hookflow signs each delivery with both
    the new secret and the retired one for the endpoint's grace window (24 hours
    by default), so the new secret can be deployed whenever you like rather than
    at the instant you press rotate. The delivery is authentic if *any* ``v1``
    matches.

    Args:
        payload: Raw request body as string
        signature: X-Signature header value (format: t=timestamp,v1=signature[,v1=...])
        secret: Endpoint webhook secret
        tolerance_ms: Maximum age of signature in milliseconds

    Returns:
        True if signature is valid

    Raises:
        HookflowError: If signature is invalid or expired
    """
    if not signature:
        raise HookflowError(
            "Missing signature header", 400, "invalid_signature"
        )

    # Parse signature
    timestamp: Optional[str] = None
    # Collected, not overwritten: a header sent during a secret rotation carries one
    # v1 per valid secret, and keeping only the last would reject whichever of the
    # pair the receiver is currently holding.
    signatures: list[str] = []

    for part in signature.split(","):
        if "=" in part:
            key, value = part.split("=", 1)
            if key == "t":
                timestamp = value.strip()
            elif key == "v1":
                signatures.append(value.strip())

    if not timestamp or not signatures:
        raise HookflowError(
            "Invalid signature format. Expected: t=timestamp,v1=signature",
            400,
            "invalid_signature",
        )

    # Check timestamp
    timestamp_ms = int(timestamp)
    now_ms = int(time.time() * 1000)

    if abs(now_ms - timestamp_ms) > tolerance_ms:
        raise HookflowError(
            "Webhook timestamp is outside tolerance window",
            400,
            "timestamp_expired",
        )

    # Verify signature
    signed_payload = f"{timestamp}.{payload}"
    expected_signature = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    # Every candidate is compared, with no early exit, so the time taken does not
    # depend on which position matched.
    matched = False
    for candidate in signatures:
        if hmac.compare_digest(candidate, expected_signature):
            matched = True

    if not matched:
        raise HookflowError("Invalid signature", 400, "invalid_signature")

    return True


def verify_standard_webhook(
    payload: str,
    headers: Dict[str, Any],
    secret: str,
    tolerance_seconds: int = DEFAULT_STANDARD_TOLERANCE_SECONDS,
) -> bool:
    """Verify the `Standard Webhooks <https://www.standardwebhooks.com>`_ headers.

    Endpoints receive both header sets by default (``signatureScheme: "BOTH"``), so use
    whichever suits you — this one if you would rather verify the same way as other
    providers you integrate with, :func:`verify_signature` if you already verify
    ``X-Signature``.

    Two things differ from Hookflow's own scheme beyond the header names: the message id is
    part of what is signed, and the digest is base64 rather than hex. Rotation behaves the
    same — through the grace window the header carries a space-separated signature per valid
    secret, and any one matching is enough.

    Args:
        payload: the raw request body.
        headers: the request headers, including the three ``webhook-*`` ones.
        secret: the endpoint's ``standardWebhooksSecret`` (``whsec_…``). A raw secret is
            accepted too and used as-is.
        tolerance_seconds: how far the timestamp may be from now, either way.

    Returns:
        True if the signature is valid.

    Raises:
        HookflowError: if it is not.
    """
    lowered = {str(k).lower(): v for k, v in headers.items()}
    message_id = lowered.get("webhook-id")
    timestamp = lowered.get("webhook-timestamp")
    signature = lowered.get("webhook-signature")

    if not message_id or not timestamp or not signature:
        raise HookflowError(
            "Missing webhook-id, webhook-timestamp or webhook-signature header",
            400,
            "invalid_signature",
        )

    try:
        timestamp_seconds = int(str(timestamp).strip())
    except ValueError:
        raise HookflowError("Invalid webhook-timestamp header", 400, "invalid_signature")

    if abs(int(time.time()) - timestamp_seconds) > tolerance_seconds:
        raise HookflowError(
            "Webhook timestamp is outside tolerance window", 400, "timestamp_expired"
        )

    # ``whsec_<base64>`` is the conventional form, and is what the endpoint's
    # standardWebhooksSecret gives you: the base64 body decodes to the key bytes. Anything
    # else is taken literally, so a raw secret still works.
    if secret.startswith("whsec_"):
        key = base64.b64decode(secret[len("whsec_") :])
    else:
        key = secret.encode("utf-8")

    signed_content = f"{message_id}.{timestamp_seconds}.{payload}".encode("utf-8")
    expected = base64.b64encode(hmac.new(key, signed_content, hashlib.sha256).digest()).decode()

    # Space-separated, one per valid secret during a rotation window. Every candidate is
    # compared with no early exit, so the time taken does not reveal which one matched.
    matched = False
    for part in str(signature).strip().split():
        version, _, candidate = part.partition(",")
        if version != "v1" or not candidate:
            continue
        if hmac.compare_digest(candidate, expected):
            matched = True

    if not matched:
        raise HookflowError("Invalid signature", 400, "invalid_signature")

    return True


def construct_event(
    payload: str,
    headers: Dict[str, str],
    secret: str,
    tolerance_ms: int = DEFAULT_TOLERANCE_MS,
) -> WebhookEvent:
    """
    Construct a webhook event from request, verifying signature.

    What Hookflow actually PUTs on the wire is the event's **payload**, not an
    envelope: a ``client.events.send(Event(type="order.completed", data={...}))``
    arrives at your endpoint as the ``data`` object alone, with the identifiers
    carried in headers (``X-Event-Id``, ``X-Delivery-Id``, ``X-Timestamp``,
    ``X-Sequence-Number``). So ``event_id`` / ``delivery_id`` / ``timestamp``
    are always populated for a real delivery and ``data`` is the parsed body,
    but ``type`` is only populated when the body itself carries a ``type`` key
    — which for a default subscription it does not. Route on the payload, or
    configure the subscription's ``payload_template`` to wrap the event so that
    ``type`` becomes part of the body.

    Args:
        payload: Raw request body as string
        headers: Request headers (case-insensitive dict)
        secret: Endpoint webhook secret
        tolerance_ms: Maximum age of signature in milliseconds

    Returns:
        Parsed and verified WebhookEvent

    Raises:
        HookflowError: If signature is invalid or payload is malformed
    """
    # Get headers (case-insensitive)
    headers_lower = {k.lower(): v for k, v in headers.items()}

    signature = headers_lower.get("x-signature", "")
    timestamp = headers_lower.get("x-timestamp", "")
    event_id = headers_lower.get("x-event-id", "")
    delivery_id = headers_lower.get("x-delivery-id", "")

    if not signature:
        raise HookflowError(
            "Missing X-Signature header", 400, "missing_header"
        )

    verify_signature(payload, signature, secret, tolerance_ms)

    # Parse payload
    import json

    try:
        data = json.loads(payload)
    except json.JSONDecodeError:
        raise HookflowError("Invalid JSON payload", 400, "invalid_payload")

    return WebhookEvent(
        event_id=event_id,
        delivery_id=delivery_id,
        timestamp=int(timestamp) if timestamp else int(time.time() * 1000),
        type=data.get("type", ""),
        data=data.get("data", data),
    )


def generate_signature(
    payload: str,
    secret: str,
    timestamp_ms: Optional[int] = None,
) -> str:
    """
    Generate a signature for testing purposes.

    Args:
        payload: Request body as string
        secret: Webhook secret
        timestamp_ms: Optional timestamp in milliseconds (defaults to now)

    Returns:
        Signature string in format t=timestamp,v1=signature
    """
    ts = timestamp_ms or int(time.time() * 1000)
    signed_payload = f"{ts}.{payload}"
    signature = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return f"t={ts},v1={signature}"
